resources/type-system/upper-lower-bounds.md

```diff
@@ -0,0 +1,390 @@
+# Dart 2.0 Upper and Lower bounds
+
+leafp@google.com
+
+This documents the currently implemented upper and lower bound computation,
+modified to account for explicit nullability and the accompanying type system
+changes (including the legacy types).  In the interest of backwards
+compatibility, it does not try to fix the various issues with the existing
+algorithm.  
+
+## Types
+
+The syntactic set of types used in this draft are a slight simplification of
+full Dart types, as described in the subtyping
+document
+[here](https://github.com/dart-lang/language/blob/master/resources/type-system/subtyping.md)
+
+## Helper predicates
+
+The **TOP** predicate is true for any type which is in the equivalence class of
+top types.
+
+- **TOP**(`T?`) is true iff **TOP**(`T`) or **OBJECT**(`T`)
+- **TOP**(`T*`) is true iff **TOP**(`T`) or **OBJECT**(`T`)
+- **TOP**(`dynamic`) is true
+- **TOP**(`void`) is true
+- **TOP**(`FutureOr<T>`) is **TOP**(T)
+- **TOP**(T) is false otherwise
+
+The **OBJECT** predicate is true for any type which is in the equivalence class
+of `Object`.
+
+
+- **OBJECT**(`Object`) is true
+- **OBJECT**(`dynamic`) is true
+- **OBJECT**(`void`) is true
+- **OBJECT**(`FutureOr<T>`) is **OBJECT**(T)
+- **OBJECT**(T) is false otherwise
+
+The **BOTTOM** predicate is true for things in the equivalence class of `Never`.
+
+- **BOTTOM**(`Never`) is true
+- **BOTTOM**(`X&T`) is true iff **BOTTOM**(`T`)
+- **BOTTOM**(`X extends T`) is true iff **BOTTOM**(`T`)
+- **BOTTOM**(`T`) is false otherwise
+
+The **NULL** predicate is true for things in the equivalence class of `Null`
+
+- **NULL**(`Null`) is true
+- **NULL**(`T?`) is true iff **NULL**(`T`) or **BOTTOM**(`T`)
+- **NULL**(`T*`) is true iff **NULL**(`T`) or **BOTTOM**(`T`)
+- **NULL**(`T`) is false otherwise
+
+The **MORETOP** predicate defines a total order on top and `Object` types.
+
+- **MORETOP**(`void`, `T`) = true
+- **MORETOP**(`T`, `void`) = false
+- **MORETOP**(`dynamic`, `T`) = true
+- **MORETOP**(`T`, `dynamic`) = false
+- **MORETOP**(`Object`, `T`) = true
+- **MORETOP**(`T`, `Object`) = false
+- **MORETOP**(`T?`, `S?`) = **MORETOP**(`T`, `S`)
+- **MORETOP**(`T?`, `S`) = true
+- **MORETOP**(`T`, `S?`) = false
+- **MORETOP**(`T*`, `S*`) = **MORETOP**(`T`, `S`)
+- **MORETOP**(`T*`, `S`) = true
+- **MORETOP**(`T`, `S*`) = false
+- **MORETOP**(`FutureOr<T>`, `FutureOr<S>`) = **MORETOP**(T, S)
+
+The **MOREBOTTOM** predicate defines a total order on bottom and `Null` types.
+
+- **MOREBOTTOM**(`Never`, `T`) = true
+- **MOREBOTTOM**(`T`, `Never`) = false
+- **MOREBOTTOM**(`Null`, `T`) = true
+- **MOREBOTTOM**(`T`, `Null`) = false
+- **MOREBOTTOM**(`T?`, `S?`) = **MOREBOTTOM**(`T`, `S`)
+- **MOREBOTTOM**(`T?`, `S`) = true
+- **MOREBOTTOM**(`T`, `S?`) = false
+- **MOREBOTTOM**(`T*`, `S*`) = **MOREBOTTOM**(`T`, `S`)
+- **MOREBOTTOM**(`T*`, `S`) = true
+- **MOREBOTTOM**(`T`, `S*`) = false
+- **MOREBOTTOM**(`X&T`, `X&S`) = **MOREBOTTOM**(`T`, `S`)
+- **MOREBOTTOM**(`X&T`, `S`) = true
```

Comment:
This is surprising; we'd have `MOREBOTTOM(X & Null, Y extends Never)` for any X and Y, which seems to allow us to pass `null` as an argument where the parameter type is actually `Never`.

Maybe we'll never have any arguments of type `Never`, but it looks like a source of unsoundness in principle.

Line 84 admits a symmetric example with the same issue.

Reply:
Well, `X&Null` is neither a bottom type nor a Null type, so that specific example doesn't arise.  I think this is another instance of the same weirdness you commented above, where in order to get a total order on Null types, I need to (or at least am choosing to) impose a total order on the union of the Null and the Never types.  

An example that could arise would be:  MOREBOTTOM(Null, Never?), which arises as the base case of the query MOREBOTTOM(Null?, Never??) . Again, we get a reasonable answer, but it's a bit weird how we get it.

An alternative way that I could formulate this and MORETOP is to explicitly handle the case that one of the arguments is BOTTOM and one is NULL.  So add rules that look like

MOREBOTTOM(T, S) = T if BOTTOM(T) and NULL(S)
MOREBOTTOM(S, T) = S if BOTTOM(S) and NULL(T)
 
I think I prefer to just using the total order, even though it's weird?

Reply:
> `X&Null` is neither

Ah, yes, sorry. The robust intuition is "a **BOTTOM** type is `Never`, spelled in different ways". The **NULL** types are a bit more tricky: `Never*` is such a type, and it could stand for `Never`:

```dart
// Opt-out.
Function(X) f<X>(Function(X) f) => f;

// Opt-in.
main() => (f<Never>((_) => true))(null);
```

So we have the dual of the top quirk: **MOREBOTTOM**(`Never`, `X & Never`), even though we generally have `X & T <: T`. Again, the most simple types are in the most extreme positions, so it's consistent.

I think it's fine to keep the rules on BOTTOM-union-NULL.